Add support for powercap sysfs

cmcantalupo opened the pull request:
- Fixes #3538
- [x] Unit tests have been mostly erased, need to add them back
- [x] CI is broken

cmcantalupo committed:
Add support for powercap sysfs

- Fixes #3438

Co-authored-by: Brad Geltz <brad.geltz@intel.com>
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on libgeopmd/test/PowercapSysfsDriverTest.cpp:
```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright (c) 2015 - 2024 Intel Corporation
```
```suggestion
 * Copyright (c) 2015 - 2025 Intel Corporation
```

avilcheslopez commented on libgeopmd/src/IOGroup.cpp:
```diff
@@ -98,6 +99,8 @@ namespace geopm
             // msr driver once it is considered more stable.
             register_plugin(CpufreqSysfsDriver::plugin_name(),
                             CpufreqSysfsDriver::make_plugin);
+            register_plugin(PowercapSysfsDriver::plugin_name(),
+                            PowercapSysfsDriver::make_plugin);
```
Question: since this plugin is going to be registered in both branches of the `if` statement (i.e., regardless of whether the user has cap sysadmin), why not have this statement just once outside of the `if`?

I do see the same pattern was followed for `CpufreqSysfsDriver`, so my question applies to both.

cmcantalupo replied:
The load order effects which aliases have priority.  The reason for the pattern is to ensure the load order (last IOGroup providing a high level alias wins).

avilcheslopez replied:
Is it that you need to make sure `PowercapSysfsDriver` is loaded after `CpufreqSysfsDriver`? If so, seeing how they're both loaded first in each of the `if` branches, you could move those two `register_plugin()` calls up right before the `if` statement, preserving the order between them and with respect to the rest of the plugins.

cmcantalupo replied:
I tried to update the comments about cpufreq vs msr support to include powercap in description.  Still should consider prefering cpufreq and powercap vs stock msr driver.

avilcheslopez commented on libgeopmd/src/PowercapSysfsDriver.hpp:
```diff
@@ -0,0 +1,37 @@
+            const std::map<std::string, properties_s> M_PROPERTIES;
+            const std::map<std::string, std::string> M_POWERCAP_RESOURCE_BY_NAME;
+            std::map<std::string, int> m_domain_map;
+            const std::string m_powercap_directory;
```
Naming convention for const member

avilcheslopez commented on libgeopmd/src/PowercapSysfsDriver.cpp:
```diff
@@ -0,0 +1,196 @@
+    std::map<std::string, std::string> result;
+
+    for (const auto &policy_file : geopm::list_directory_files(powercap_directory)) {
+        if (policy_file.find("intel-rapl:") != 0) {
+            continue;
+        }
+
+        std::ostringstream oss;
+        oss << powercap_directory << "/" << policy_file << "/name";
+        auto name_path = oss.str();
+
+        int name_fd = open(name_path.c_str(), O_RDONLY);
+        if (name_fd == -1) {
+            throw geopm::Exception("PowercapSysfsDriver failed to open " + name_path,
+                                   errno, __FILE__, __LINE__);
+        }
+        char name_buf[geopm::SysfsDriver::M_IO_BUFFER_SIZE] = {};
+        int read_bytes = pread(name_fd, name_buf, sizeof(name_buf) - 1, 0);
+        close(name_fd);
+        if (read_bytes < 0) {
+            throw geopm::Exception("PowercapSysfsDriver failed to read " + name_path,
+                                   errno, __FILE__, __LINE__);
+        }
+        if (read_bytes == sizeof(name_buf) - 1) {
+            throw geopm::Exception("PowercapSysfsDriver truncated read from " + name_path,
+                                   errno, __FILE__, __LINE__);
+        }
```
I'm seeing there is opportunity to factor out common code with `CpufreqSysfsDriver`, since I'm seeing the good majority of the code in this function is practically the same. At the very least, this section of code could be factored out into a helper function somewhere that both drivers could leverage.

cmcantalupo replied:
Although there are similarities between load_powercap_resource_by_name() and load_cpufreq_cpus_by_resource(), there are some key differences.  The code that is common, is that we are opening all of the files in a directory, reading the contents and doing something depending on the contents.  Much of the work to do that is already factored in geopm::list_directory_files().  

The key differences are the use of the "name" file in the case of powercap, and "affected_cpus" in the cpufreq case to understand the topology.  These file contents are quite different.  I don't see a really easy way to share the code without using std::functions() to be applied to the contents of the files, but that seems overly complex and difficult to read.

avilcheslopez commented on libgeopmd/src/PowercapSysfsDriver.cpp:
```diff
@@ -0,0 +1,196 @@
+        return GEOPM_DOMAIN_PACKAGE;
+    }
+
+    static std::string name_to_resource_key(const std::string &name, int domain_idx)
```
Minor recommendation: could these static functions be moved to the top (to facilitate reading the code)? It might just throw someone off a bit when they're going through class functions, and then encounter a static function not part of the class. It might just be me, so feel free to ignore this suggestion.

avilcheslopez commented on libgeopmd/src/PowercapSysfsDriver.cpp:
```diff
@@ -0,0 +1,196 @@
+    static std::string name_to_property_key(const std::string &name)
+    {
+        std::string property_key;
+        auto substr_pos = name.find("POWERCAP::CPU_");
+        if (substr_pos == std::string::npos) {
+            substr_pos = name.find("POWERCAP::DRAM_");
```
Minor suggestion: seeing these literal strings (passed to the `find` function) are used in more than one place, I would recommend putting them in a constant.

avilcheslopez commented on libgeopmd/test/PowercapSysfsDriverTest.cpp:
```diff
@@ -0,0 +1,200 @@
+    }
+}
+
+TEST_F(PowercapSysfsDriverTest, driver_and_plugin_name_match)
+{
+    EXPECT_EQ("POWERCAP", m_driver->driver())
+        << "Driver name should match the plugin name";
+    EXPECT_EQ("POWERCAP", PowercapSysfsDriver::plugin_name())
+        << "Plugin name should be POWERCAP";
+}
```
It seems this is a duplicate of the first test (`iogroup_plugin_name_matches_driver_name`).

avilcheslopez commented on libgeopmd/src/PowercapSysfsDriver.cpp:
```diff
@@ -0,0 +1,196 @@
+        char name_buf[geopm::SysfsDriver::M_IO_BUFFER_SIZE] = {};
+        int read_bytes = pread(name_fd, name_buf, sizeof(name_buf) - 1, 0);
+        close(name_fd);
+        if (read_bytes < 0) {
```
Is it possible that `pread` in this case could ever return 0 bytes? I'm guessing the answer is highly unlikely, but perhaps we should manage that scenario just in case.

cmcantalupo replied:
"The name_buf array is initialized to all zeroes. This is because the {} initializer in C++ initializes the array elements to their default value, and for a char type, the default value is 0 (null character '\0')."

- Copilot

So in the case of read_bytes == 0, the name_buf should begin with '\0' and thus be converted to an empty std::string name below

cmcantalupo committed:
Some code clean up

- Const member in all caps
- Move static functions out of geopm namespace and to top of file
- Remove duplicate test fixture
- Improve comments about plugin load order
- Define static const for repeated strings

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented on libgeopmd/src/IOGroup.cpp:
```diff
@@ -94,10 +95,14 @@ namespace geopm
-            // May want to give this higher priority than the non-safe
-            // msr driver once it is considered more stable.
+            // May want to give Cpufreq and Powercap higher priority than the
+            // non-safe msr driver once it is considered more stable.  The
```
```suggestion
            // non-safe msr driver once they are considered more stable.  The
```

cmcantalupo committed:
Fix tense in comment
